distributed/worker.py

```diff
@@ -3444,47 +3460,97 @@ async def find_missing(self) -> None:
+                        MissingDataMsg(
+                            key=key, errant_worker=self.address, stimulus_id=stimulus_id
                         )
-                        # Do not mutate the input dict. That's rude
-                        workers = set(workers) - {self.address}
-                    dep_ts.who_has.update(workers)
+                    )
 
-                    for worker in workers:
-                        self.has_what[worker].add(dep)
-                        self.data_needed_per_worker[worker].push(dep_ts)
-        except Exception as e:  # pragma: no cover
-            logger.exception(e)
-            if LOG_PDB:
-                import pdb
+            if ts.who_has == workers:
+                continue
 
-                pdb.set_trace()
-            raise
+            self.log.append(
+                (
+                    key,
+                    "update-who-has",
+                    list(ts.who_has),
+                    list(workers),
+                    stimulus_id,
+                    time(),
+                )
+            )
+
+            for worker in ts.who_has - workers:
+                self.has_what[worker].discard(key)
+                # Can't remove from self.data_needed_per_worker; there is logic
+                # in _select_keys_for_gather to deal with this
+
+            for worker in workers - ts.who_has:
+                self.has_what[worker].add(key)
+                if ts.state == "fetch":
+                    self.data_needed_per_worker[worker].push(ts)
+                    # All workers which previously held a replica of the key may either
+                    # be in flight or busy. Kick off ensure_communicating to try
+                    # fetching the data from the new worker.
+                    # There are other reasons why a task may be sitting in 'fetch' state
+                    # - e.g. if we're over the total_out_connections or the
+                    # comm_threshold_bytes limit, or if we're paused. We're deliberately
+                    # NOT testing the full gamut of use cases for the sake of simplicity
+                    # and robustness and just stating that ensure_communicating *may*
+                    # return new GatherDep events now.
+                    instructions.append(
+                        EnsureCommunicatingAfterTransitions(stimulus_id=stimulus_id)
+                    )
+
+            ts.who_has = workers
+            # currently fetching -> can no longer be fetched -> transition to missing
+            # any other state -> eventually, possibly, the task may transition to fetch
+            # or missing, at which point the relevant transitions will test who_has that
+            # we just updated. e.g. see the various transitions to fetch, which
+            # instead recommend transitioning to missing if who_has is empty.
+            if not workers and ts.state == "fetch":
+                recs[ts] = "missing"
```

Comment:
A major part of the recent refactoring is to separate comm handlers from state machine mutations by introducing StateMachineEvents as an interemdiate interface. Why is this not issuing state machine events? I don't like us introducing more recommendations.

This also causes us to implement the rather artificial transitions + handle_instructions calls that were already mentioned in https://github.com/dask/distributed/pull/6342/files#r874029445

Reply:
Zooming out a bit, the two "events" I can see here are

- `NewReplicasOnCluster`
- `LostReplicaOnCluster`

In this method we're somehow entangling the generation of these events and how they should be handled.

Reply:
> NewReplicasOnCluster
> LostReplicaOnCluster

These aren't events; this is information that is carried by different events.
The events are three; ``_update_who_has`` is a helper function called by them:

- ``handle_compute_task`` (to be trivially refactored soon as ``ComputeTaskEvent``)
- ``handle_acquire_replicas`` (to be trivially refactored soon as ``AcquireReplicasEvent``)
- ``RefreshWhoHasEvent`` (#6348), which in turn is triggered by
  - ``find_missing`` -> ``RequestRefreshWhoHasMsg`` (#6348)
  - ``gather_dep`` -> ``refresh_who_has`` -> ``RequestRefreshWhoHasMsg`` (#6348)

Reply:
> * I believe the **who_has update** in `handle_compute_task` and `handle_acquire_replicas` should never actually trigger any other transition the compute-task/acquire-replica call wouldn't generate anyhow, should they?

`handle_compute_task` and `handle_acquire_replicas` can't cause transitions to missing.
However, if

1. a key is sitting in fetch state because all known workers are in flight or busy
2. scheduler calls handle_compute_task or handle_acquire_replicas and it incudes that key in who_has
3. a new worker is discovered which holds a replica
4. the key is immediately transitioned to flight.

In main, the key will sit in fetch state until something else - namely, those workers going out of flight/busy state - kicks off ensure_communicating.